teps/0156-whenexpressions-in-step.md

```diff
@@ -0,0 +1,235 @@
+**Examples**
+ - Operator-based when expressions in steps.
+```yaml
+  apiVersion: tekton.dev/v1
+  kind: TaskRun
+  metadata:
+    generateName: operator-based-
+  spec:
+    workspaces:
+      - name: custom
+        persistentVolumeClaim:
+        claimName: my-pvc-2
+    taskSpec: 
+      steps:
+        - name: produce-step
+          image: alpine
+          results:
+             - name: result2
+               type: string
+          script: |
+             echo -n "foo" | tee $(step.results.result2.path)
+        - name: run-based-on-step-results
+          image: alpine
+          script: |
+             echo "wooooooo"
+          when:
+             - input: "$(steps.produce-step.results.result2)"
+               operator: in
+               values: ["bar"]   
+```
+
+- CEL when expressions in steps.
+
+```yaml
+  apiVersion: tekton.dev/v1
+  kind: TaskRun
+  metadata:
+     generateName: cel-based-
+  spec:
+    taskSpec:
+      steps:
+        - name: skipped-step-cel
+          image: alpine
+          when:
+            - cel: "'monday'=='friday'"
+          script: |
+            echo -n "should not see me"
+```
+### Example Usage
+Here are some use cases that showcase the power of this feature:
+
+**1. Caching with Conditional Steps:**
+
+Imagine a Task that performs complex calculations. To improve performance, you want to implement caching but only when certain conditions are met:
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: my-complex-task
+spec:
+  steps:
+    - name: check-cache
+      image: alpine
+      results:
+        - name: cacheHit
+          type: string
+      script: |
+        ## some logic to check cache and fetch if cache is available and set cacheHit to true
+        echo -n "true" >> $(step.results.cacheHit.path)
+    - name: calculate-results
+      image: alpine
+      when:
+        - input: "$(steps.check-cache.results.cacheHit)"
+          operator: notin
+          values: ["true"]
+      script: |
+        echo "Performing complex calculations"
+```
+
+In this example, if the `CacheHit` is set to "true" in the `check-cache` step the `calculate-results` step will be skipped.
+
+**2. Reporting and Failure Handling:**
+
+You might want to add a reporting step that summarizes the outcome of a previous step, only when that step fails:
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: my-task-handling-timeout
+spec:
+  steps:
+    - name: main-task
+      image: alpine
+      script: |
+        echo 123
+    - name: report-timeout
+      image: alpine
+      when:
+        - input: "$(steps.main-task.status.terminationReason)"
+          operator: in
+          values: ["TimeoutExceeded"]
+      script: |
+        echo "reporting timeout!"
+```
+
+Here, the `report-timeout` step will only run if the `main-task` step terminates with a "TimeoutExceeded" status.  This allows you to specifically handle failures without executing the reporting logic every time.
```

Comment:
In that case, the `Task` would still be marked as a failure, isn't it ?
If yes, is there a case where we wouldn't want that behavior ?

Reply:
One potential use case could be a task that performs multiple integration test via separate stepActions for each integration service. We may want to file a bug report for each integration failure, mainly because those will be investigate by different teams. Report filing might be another step Action (e.g. jira-open-bug). We can declaratively control bug filing for each integration run via `when` expression.

Reply:
I just changed this example a little bit, it turns out with the existing implementation, timeout is a error that will always fail subsequent steps, https://github.com/tektoncd/pipeline/blob/9c315909a373a600eb42f343584badeea818607c/pkg/entrypoint/entrypointer.go#L255-L257 . 

Task authors should use "onError:continue" on the steps that could potentially fail to allow the follow-up steps to execute, authors can decide if the task should fail on the step handling the error.